src/openlcb/DatagramCan.cxx

```diff
@@ -77,13 +79,37 @@ class CanDatagramClient : public DatagramClient,
         }
         HASSERT(b->data()->mti == Defs::MTI_DATAGRAM);
         result_ = OPERATION_PENDING;
-        register_handlers();
         reset_message(b, priority);
+        start_flow(STATE(acquire_srcdst_lock));
+    }
+
+    Action acquire_srcdst_lock()
+    {
+        // First check if there is another datagram client sending a datagram
+        // to the same target node.
+        {
+            AtomicHolder h(&LinkedObject<CanDatagramClient>::headMu_);
+            for (CanDatagramClient* c = LinkedObject<CanDatagramClient>::head_;
+                 c;
+                 c = c->LinkedObject<CanDatagramClient>::link_next()) {
+                // this will catch c == this.
+                if (!c->sendPending_) continue;
+                if (c->nmsg()->src.id != nmsg()->src.id) continue; 
+                if (!async_if()->matching_node(c->nmsg()->dst, nmsg()->dst))
+                    continue;
+                // Now: there is another datagram client sending a datagram to
+                // this destination. We need to wait for that transaction to
+                // complete.
+                c->waitingClients_.push_front(this);
+                return wait();
+            }
```

Comment:
I wonder if you want to set sendPending_ before you leave the AtomicHolder context.  I don't fully understand the threading model here, but it seems to me that you might have a race condition.

Reply:
For readability I would want to keep modifications to sendPending_ inside the register_handler and unregister_handler functions. This way it is easy to understand the contract.
The race condition is benign so long as all the datagram clients are running on the same service executor, since then the executor thread will prevent them from running in parallel. The atomic holder here is needed more for the iteration over the linkedobject list, because it is not guaranteed that the datagram clients are also *created* on the main executor thread.
I will try to move the entire register_handlers() call inside the lock and see if that is possible. (There is a limit to what operations we are allowed to do inside a critical section; a mutex acquisition is not allowed. The handler registry might be using mutexes internally, so it's not clear that this is safe.)

Reply:
I think that threading model makes sense.  I assumed the Atomic is used in order to avoid the overhead of a mutex.  If a mutex is a possibility without adding much overhead (reuse an existing mutex for example), it might be preferred since technically the list size is unbounded.  Though practically speaking, the list size should not get that big.

Reply:
I checked and it can't be moved inside the atomic, because the dispatcher register_handler uses an OSMutex inside.
In practice this list is exactly two entries long for all freertos targets.